File: src/main/java/ru/chernov/urlshortener/validation/constraint/AvailableUsername.java
package ru.chernov.urlshortener.validation.constraint;

import jakarta.validation.Constraint;
import jakarta.validation.Payload;
import ru.chernov.urlshortener.validation.validator.AvailableUsernameValidator;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;


@Target({ElementType.FIELD})
@Retention(RetentionPolicy.RUNTIME)
@Constraint(validatedBy = AvailableUsernameValidator.class)
public @interface AvailableUsername {

    String message() default "Username already exists";

    Class<?>[] groups() default {};

    Class<? extends Payload>[] payload() default {};

}

File: src/main/kotlin/ru/chernov/urlshortener/validation/constraint/AvailableUsername.kt
package ru.chernov.urlshortener.validation.constraint

import jakarta.validation.Constraint
import jakarta.validation.Payload
import ru.chernov.urlshortener.validation.validator.AvailableUsernameValidator
import kotlin.reflect.KClass


@Target(AnnotationTarget.FIELD)
@Retention(AnnotationRetention.RUNTIME)
@Constraint(validatedBy = [AvailableUsernameValidator::class])
annotation class AvailableUsername(
        val message: String = "Username already exists",
        val groups: Array<KClass<*>> = [],
        val payload: Array<KClass<out Payload>> = []
)